File: nym-vpn-android/vpn-client/src/main/java/net/nymtech/vpn/model/EntryPoint.kt
package net.nymtech.vpn.model

sealed class EntryPoint {
    //TODO enforce only two char countryISO
    data class Location(private val location: String) : EntryPoint() {

        //TODO make this serialize later
        override fun toString(): String {
            return "{ \"Location\": { \"location\": \"${location}\" }}"
        }
    }
    private sealed class Gateway() : EntryPoint() {

        //TODO example, impl later
        override fun toString(): String {
            return "{ \"Gateway\": { \"identity\": [94,69,76,90,128,87,76,174,15,177,79,44,11,234,27,225,205,162,191,216,144,29,74,210,50,62,121,13,154,85,209,40] }}"
        }
    }
}

File: nym-vpn-core/crates/nym-gateway-directory/src/entries/entry_point.rs
// Copyright 2023 - Nym Technologies SA <contact@nymtech.net>
// SPDX-License-Identifier: GPL-3.0-only

use std::fmt::{Display, Formatter};

use nym_sdk::mixnet::NodeIdentity;
use serde::{Deserialize, Serialize};

use crate::{Error, error::Result};

// The entry point is always a gateway identity, or some other entry that can be resolved to a
// gateway identity.
#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
pub enum EntryPoint {
    // An explicit entry gateway identity.
    Gateway { identity: NodeIdentity },
    // Select a random entry gateway in a specific country.
    Country { two_letter_iso_country_code: String },
    // Select a random entry gateway in a specific region/state.
    Region { region: String },
    // Select an entry gateway at random.
    Random,
}

impl Display for EntryPoint {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            EntryPoint::Gateway { identity } => write!(f, "Gateway: {identity}"),
            EntryPoint::Country {
                two_letter_iso_country_code,
            } => write!(f, "Country: {two_letter_iso_country_code}"),
            EntryPoint::Region { region } => write!(f, "Region/state: {region}"),
            EntryPoint::Random => write!(f, "Random"),
        }
    }
}

impl EntryPoint {
    pub fn from_base58_string(base58: &str) -> Result<Self> {
        let identity = NodeIdentity::from_base58_string(base58).map_err(|source| {
            Error::NodeIdentityFormattingError {
                identity: base58.to_string(),
                source,
            }
        })?;
        Ok(EntryPoint::Gateway { identity })
    }
}
